Mathlib/MeasureTheory/Group/ModularCharacter.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2025 Noam Atar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Noam Atar
+-/
+import Mathlib.MeasureTheory.Function.LocallyIntegrable
+import Mathlib.MeasureTheory.Group.Integral
+import Mathlib.MeasureTheory.Group.Measure
+import Mathlib.Topology.Metrizable.Urysohn
+import Mathlib.MeasureTheory.Measure.Haar.Unique
+
+
+/-!
+# Modular character of a locally compact group
+
+On a locally compact group, there is a natural character G → ℝ*, which for g : G gives the value
+μ (· * g) / μ, where μ is an (inner regular) Haar measure. This file defines this character, called
+the modular character, and shows that it is multiplicative and independent of the chosen Haar
+measure.
+TODO: Show that the character is continuous.
+
+## Main Declarations
+
+* `modularCharacter_fun`: Given a measure μ, define the modular character associated with this
+  measure, using the (noncomputable) `haarScalarFactor`. The result that this does not depend on the
+  measure chosen (as long as it is inner regular) is `modularCharacter_fun_eq`.
+* `modularCharacter`: The homomorphism G →* ℝ≥0 whose toFun is `modularCharacter_fun`.
+-/
+
+open MeasureTheory
+open scoped NNReal
+
+namespace MeasureTheory
+
+namespace Measure
+
+variable {G : Type*} [TopologicalSpace G] [Group G] [IsTopologicalGroup G]
+  [MeasurableSpace G] [BorelSpace G] [LocallyCompactSpace G]
+
+open Classical
+
+/-- Definition of the modularCharacter map -/
+@[to_additive addModularCharacterFun "Definition of the addModularCharacter map"]
+noncomputable def modularCharacterFun : G → ℝ≥0 :=
+  fun g => haarScalarFactor (map (· * g) MeasureTheory.Measure.haar) MeasureTheory.Measure.haar
```

Comment:
You shouldn't assume that `G` is a Borel measurable space for the definition: the definition makes sense if `G` is a locally compact group, and you should endow `G` with its Borel structure inside the definition.

Reply:
Could you explain how to do this?

Reply:
Something like
```
noncomputable def modularCharacterFun (g : G) : ℝ≥0 :=
  letI : MeasurableSpace G := borel G
  haveI : BorelSpace G := ⟨rfl⟩
  haarScalarFactor (map (· * g) MeasureTheory.Measure.haar) MeasureTheory.Measure.haar
```
should work.

Reply:
It seems like it works, but then `haarScalarFactor` can't be initialized, for example in `modularCharacterFun_eq_haarScalarFactor`.

Reply:
In other lemmas that mention a measure, you will need to assume `[MeasurableSpace G] [BorelSpace G]` so that the notion of measure makes sense. But in lemmas that don't mention a measure, you won't need it.

Reply:
Then I have to apply some uniqueness result about `BorelSpace`. For example now `modularCharacterFun g = haarScalarFactor (map (· * g) haar) haar` cannot be proven by `rfl`. Do you know how to do this?

Reply:
You can prove it with `:= by borelize G; rfl`

Reply:
Done. Thank you!